glean-core/src/database/mod.rs

```diff
@@ -16,6 +19,21 @@ use crate::Glean;
 use crate::Lifetime;
 use crate::Result;
 
+lazy_static!{
+    // A map of all the known 'user' lifetime metrics along with their expiration
+    // dates. See bug 1604854 for more context on why this is needed.
+    static ref USER_LIFETIME_EXPIRATION_MAP: HashMap<&'static str, NaiveDate> = vec![
+        ("glean_internal_test.user_metric_expired", NaiveDate::from_ymd(2015, 3, 14)),
+    ].into_iter().collect();
+}
+
+lazy_static!{
+    // Get the current date at runtime. While we should use the build-date to be
+    // consistent with other metric types, given the edge case this already is,
+    // it might be enough to go with a run-time date.
+    static ref STARTUP_LOCAL_DATE: NaiveDate = Local::now().naive_local().date();
```

Comment:
@badboy @mdboom what's your take on this simplification? If we were to go for the "build time" date, we'd need to change the FFI to pass in a date from e.g. the Android build system. Given the fact that this is an edge case already, we might as well simplify our lives.

Reply:
I think this is fine, but maybe either (1) use UTC here (which is what metric expiration at build time is based on), or (2) add 36 hours so we would never remove metrics before the expected build time.  (2) may actually be better to account for clock skew.

Reply:
That is ... certainly a change in what we communicated previously. I'm fine with that, but we should make that clear.
Multiple questions I have:

* This date is only set on first access. Should we make sure to init it in initialize?
* Do we already have a list of expired/expiring user lifetime metrics?
	* That would allow us to evaluate the following: Can we just build Glean at a point where all those are expired and therefore hard-code their deletion?
	* Are we going to disallow expiration of user-lifetime metrics via glean_parser?

Reply:
> That is ... certainly a change in what we communicated previously.

You mean the UTC part?  I think that's been the definition from the beginning (so that builds built in multiple locations wouldn't be different)  https://mozilla.github.io/glean_parser/metrics-yaml.html#expires

Reply:
No the "datetime at runtime" vs "build time"

Reply:
> 
> 
> That is ... certainly a change in what we communicated previously. I'm fine with that, but we should make that clear.
> Multiple questions I have:
> 
>     * This date is only set on first access. Should we make sure to init it in initialize?

Good point.

>     * Do we already have a list of expired/expiring user lifetime metrics?

I'm building this list, it will come as a separate commit.

>       * That would allow us to evaluate the following: Can we just build Glean at a point where all those are expired and therefore hard-code their deletion?

To make sure I understand, @badboy , do you mean postponing this change to a "future date" in which we'll just release the code to clear them all up?

>       * Are we going to disallow expiration of user-lifetime metrics via glean_parser?

I filed [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1626516) for that.

Reply:
@mdboom what will the 36 hours offset give us?

Reply:
> do you mean postponing this change to a "future date" in which we'll just release the code to clear them all up?

Yeah. If we know we have expired ones, why not clean only those up when we know that's the case. We don't allow expiry in the future so we're good.

Reply:
> 
> 
> > do you mean postponing this change to a "future date" in which we'll just release the code to clear them all up?
> 
> Yeah. If we know we have expired ones, why not clean only those up when we know that's the case. We don't allow expiry in the future so we're good.

That's another strategy, I'd be fine with that. It puts the burden of remembering when things expire on us, though.

Reply:
It also requires Fenix (or similar) to upgrade Glean in a timely manner after the expiration.
Still I think it's a worthwhile way to discuss.

Reply:
> @mdboom what will the 36 hours offset give us?

Given clock skew and clocks set in the entirely wrong timezones, an assurance that user metrics are never deleted early.  But on second thought, this is just too much of a fudge.  Using UTC rather than local seems like low-hanging fruit, but let's not add any additional weirdness on top.

Reply:
@mdboom , what's your take on what @badboy suggested a couple of comments above?

Reply:
If we could take all date parsing at runtime out and just assume metrics don't go in the remove list until they are already expired, it does simplify things.  We will have to remember to do it, but maybe we set a calendar reminder (for ourselves) to update that list when we know they are going to expire?

Reply:
Apologies for keeping the conversation going, but I still have a few doubts about the next step. As far as I understood, the discussion above distills down to a few options.

1. Merge this PR as it is, add the expired metrics later when they are about to expire. We'd still need to coordinate and cut a release before metrics expire.
2. Drop this PR, resume it *after* the metrics we're looking to expire are actually expired, then cut a new release of the SDK that will purge that data in a way similar to the one in this PR.
3. Merge this PR together with a commit with the expiration dates.
4. Do nothing, postpone decision to when metrics are expired.

@badboy @mdboom I don't think (1) gives us any advantage over (4). I'd be happy with either (2) or (3) happening. What's your take?

Reply:
So ... I did take a brief look into Fenix and GeckoView for expiring user-lifetime metrics:

* `activation.activation_id` expires 2020-09-01
* a bunch of metrics in m-c: expire 2021-02-01

Am I missing other metrics we know about?
That at least gives us ample time to tackle this, obviously I'd like this to move sooner.

The code approach here looks good to me.

I think the only thing I'd like to see a decision on is whether we base this on a runtime date as the PR currently does or on the build-time date of Glean (with a reminder for us to push releases on those two dates above).
(I'm explicitly ignoring a "pass in build time of app" approach, that's more hassle than it's worth).

Does that make sense?

Reply:
> 
> 
> So ... I did take a brief look into Fenix and GeckoView for expiring user-lifetime metrics:
> 
>     * `activation.activation_id` expires 2020-09-01
> 
>     * a bunch of metrics in m-c: expire 2021-02-01
> 
> 
> Am I missing other metrics we know about?

Not that I'm aware of, no.

> I think the only thing I'd like to see a decision on is whether we base this on a runtime date as the PR currently does or on the build-time date of Glean (with a reminder for us to push releases on those two dates above).

I have a slight preference on "runtime date". While I know this makes it less deterministic, as it will heavily depend on the time the application runs, it gives us a guaranteed lower bound date (the expiration date) before which nothing gets removed. After that date we should not rely on the data anyway, so users of the data will be aware.

Using the SDK build date will provide us with a strict deterministic date for expiration, but given the time it will take for the SDK to propagate to A-C and then Fenix, I think the result will be pretty much the same.